cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ApplicationLifecycleService.java

```diff
@@ -345,6 +362,160 @@ public ApplicationWithPrograms updateApp(ApplicationId appId, AppRequest appRequ
                      programTerminator, ownerAdmin.getOwner(appId), appRequest.canUpdateSchedules());
   }
 
+  /**
+   * Upgrades an existing application by upgrading appliation artifact versions and plugin artifact versions.
+   *
+   * @param appId the id of the application to upgrade.
+   * @param programTerminator a program terminator that will stop programs that are removed when updating an app.
+   *                          For example, if an update removes a flow, the terminator defines how to stop that flow.
+   * @return Details of upgrade for the application like failure details with reason.
+   * @return
+   * @throws Exception
+   */
+  public ApplicationUpdateDetails upgradeApplication(ApplicationId appId, ProgramTerminator programTerminator)
+      throws Exception {
```

Comment:
indentation

Reply:
it's basically never a good idea to throw a generic Exception in an internal method, as it makes it impossible for the caller to handle errors properly. Can this throw more specific exceptions? 

We have a lot of bad existing code that does this, so if it may not be possible right now,
but you should at least add javadocs about the more specific exceptions that you know it can throw.

Reply:
Added a few that I could trace down.